backend/server/src/derivation.rs

```diff
@@ -0,0 +1,142 @@
+    /// trying to build the same derivation again. If it failed once, it will always fail.
+    ///
+    /// This counter is intended for cases in which the derivation build was interrupted due to
+    /// external factors (see [DrvBuildState::Interrupted]). In these situations it may make sense
+    /// to reattempt the build (depending on the interruption kind).
+    pub build_attempt: NonZero<u32>,
+}
+
+/// Metadata about a derivation build.
+///
+/// This metadata is useful to reproduce a build on a different machine. Note that in general,
+/// only builds that ended with a state of [DrvBuildState::Completed] can be reproduced.
+pub struct DrvBuildMetadata {
+    /// The derivation build this metadata is associated with.
+    pub build: DrvBuildId,
+
+    /// The Git commit this derivation build is based upon.
+    ///
+    /// Note that this may not be the only commit that can produce this derivation. Because a
+    /// derivation only needs to fully build once, later commits may still include this
+    /// derivation but do not trigger a new build.
+    pub commit: gix_hash::ObjectId,
+
+    /// The Nix command that was used to build this derivation.
+    pub build_command: DrvBuildCommand,
+}
+
+/// Command used to build the derivation.
+pub enum DrvBuildCommand {
+    /// Nix 1.0 style command. Fields are TBD.
+    Nix1,
+}
+
+/// Build events are emitted whenever a derivation build's state changes.
+pub struct DrvBuildEvent {
+    /// The derivation build this event is associated with.
+    pub build: DrvBuildId,
+
+    /// Monotonically increasing numeric event identifier. Use this to determine the order of
+    /// events. Events with a higher id happened later.
+    ///
+    /// Since this id is global it can be used to retrieve the latest build state for a derivation
+    /// without having to check [DrvBuild::build_attempt].
+    pub event_id: u32,
+
+    /// The build state this event propagates.
+    pub state: DrvBuildState,
+
+    /// The timestamp when this event happened.
+    ///
+    /// This timestamp only has second accuracy, which makes it unsuitable for sorting of build
+    /// events. If for example the build queue is empty, it is not unlikely that a build is
+    /// scheduled ([DrvBuildState::Pending]) and started ([DrvBuildState::Building]) within the
+    /// same second.
+    ///
+    /// Using a higher accuracy timestamp would make it incompatible with SQlite's datetime
+    /// functions which is also not desirable.
+    ///
+    /// Instead, use [event_id] to determine the order of events.
+    pub timestamp: chrono::DateTime<chrono::Utc>,
+}
+
+/// Describes the possible states a derivation build can be in.
+pub enum DrvBuildState {
+    /// Build is scheduled in a queue and has not yet started.
+    Pending,
+    /// Build is currently running.
+    Building,
+    /// Build has completed, either successfully or not.
+    Completed {
+        result: DrvBuildResult,
```

Comment:
to simplify the data model a bit
```suggestion
        was_successful: bool,
```

Reply:
I originally anticipated different data for the ok and error variants but I have since come to the realisation that doing that would make mapping the model to the database quite difficult. So I agree. We can change it back later if necessary.

Reply:
Decided to keep the enum as it is more expressive than a single boolean. Added some methods to it for turning it into a bool instead. See 8c3e740.